libs/sh1107/sh1107.cpp

```diff
@@ -126,18 +137,24 @@ void Sh1107::flush()
 
         buffer += height;
     }
+    m_current_segment++;
 }
 
 void Sh1107::draw_pixel(uint8_t x, uint8_t y)
 {
-    if (x > width || y > height)
+    if (x < 64 * m_current_segment || x >= 64 * (m_current_segment + 1) || y > height)
```

Comment:
I get your point. I had a similar thought when implementing it, but I wanted to minimize the impact on the ui lib. 

I do like my suggestion, regarding the cleanliness and readability of `ui.update()`. Reason: In my case (`sh1107`), I segmented it into the left and right half of the display. This, however, may be different for different displays as addressing for those may be more convenient by segmenting vertically. Hence, we would have to consider vertical as well as horizontal segmentation in all coordinates provided in `ui.update()`, rendering it messy and less readable. 

After all, we never really draw on a pixel level in the ui but use `draw_bitmap` instead. Therefore, I suggested to bail out at the beginning of `draw_bitmap`, if it is entirely out of the current segment ([see `sh1107.cpp`, ll. 153-157](https://github.com/brewpeople/brewslave/pull/29/commits/9756b31a7af3d0e12395518d50dd7fe05c916b49#diff-1c5aa20101e05bc8863dc003f3e7b344e8deeb48e4c8e67fe817af88d34c7d97R153-R157)). Never the less, pixels from bitmaps either entirely or partially in the current segment are still checked individually.

I could further reduce the load doing something like the following in the respective display's lib:
```
void draw_bitmap() {
    if (bitmap entirely outside of current segment) { // this is what I already suggested
        return;
    }

    if (bitmap entirely inside of current segment) {
        // draw pixels without checking, using new private functions `draw_bitmap_unchecked()` and `draw_pixel_unchecked`,
        // because it was already ensured that the bitmap is entirely inside the current segment
        draw_bitmap_unchecked(...);
    }
    else {
        // here, I could check which part of the bitmap is in the current segment
        // and draw only the overlapping portion 
        draw_bitmap_unchecked(limited to overlapping coordinates between current segment and bitmap);
    }
}
```
I'll make a suggestion.

Reply:
@matze have a look at the latest commit c51ea99.

I limited the check to the bitmap level introducing a private `draw_pixel_unchecked()`.